packages/nextjs/hooks/scaffold-eth/contract.types.ts

```diff
@@ -0,0 +1,41 @@
+import { Abi, AbiParametersToPrimitiveTypes, ExtractAbiEvent, ExtractAbiEventNames, ExtractAbiFunction } from "abitype";
+import contracts from "../../generated/hardhat_contracts";
+
+export type DefaultChain = "31337";
+
+export type Chain = keyof typeof contracts;
+
+type Contracts = typeof contracts[DefaultChain][0]["contracts"];
+
+export type ContractName = keyof Contracts;
+
+export type Contract<TContractName extends ContractName> = Contracts[TContractName];
+
+type InferContractAbi<TContract> = TContract extends { abi: infer O } ? O : never;
```

Comment:
I tried to understand why is zero here. What is it does? Probably it's better to change that `O` type

Reply:
It's just a generic type that infers it's type from the abi. I'll improve the name a bit.